### PR TITLE
More robust Sankey

### PR DESCRIPTION
**Pull request by alexcjohnson**
Make Sankey more accepting of messy data (fixing #2480)
- Numeric strings are OK
- Zero (and negative) link values get ignored completely. Note: this is not what I proposed in https://github.com/plotly/plotly.js/issues/2480#issuecomment-387201484 - thinking a bit more about it, links with no value should really be not links at all, and nodes with no size shouldn't be nodes at all. If you really want a node & link to show up with zero value, create one with an explicitly tiny value.
- Unused nodes and malformed links are ignored as well. Since there are now various places this can happen and in other contexts we simply ignore partially-invalid data points, I also removed the logged error on missing nodes - which happened during `supplyDefaults`, so was called frequently, and had a fairly slow implementation so could be a drag on large Sankey traces.

cc @etpinard

### CODE REVIEW

File: src/plot_api/plot_schema.js [hunk 1]
```diff
@@ -437,6 +437,8 @@ function recurseIntoValObject(valObject, parts, i) {
     return valObject;
 }
 
+// note: this is different from Lib.isIndex, this one doesn't accept numeric
+// strings, only actual numbers.
```
Comment (alexcjohnson) [inline src/plot_api/plot_schema.js]: I initially made `isIndex` into a `Lib` function thinking I could reuse it here... but in this context it's in principle important that an index is actually a number, to distinguish an array index from an object property that happens to be a number. That said I don't think it's possible to use a number as an object property in `nestedProperty`...

File: src/traces/sankey/render.js [hunk 2]
```diff
@@ -63,44 +66,79 @@ function switchToSankeyFormat(nodes) {
...
+        else removedNodes = true;
+    }
+
+    // need to re-index links now, since we didn't put all the nodes in
+    if(removedNodes) {
+        for(i = 0; i < links.length; i++) {
+            links[i].source = nodeIndices[links[i].source];
+            links[i].target = nodeIndices[links[i].target];
+        }
+    }
```
Comment (alexcjohnson) [inline src/traces/sankey/render.js]: For the most part it didn't cause problems to have extra unused nodes in the list, but sometimes the layout algorithm would reserve extra space for these nodes (leaving extra margins or making the whole diagram S-shaped, for example) so it's better to remove them. The `sankey_messy` mock includes a whole bunch of extra nodes both before (to test this link reassignment) and after the real ones to test this.

File: src/plot_api/plot_schema.js (hunk 1 above)
Comment (etpinard) [reply src/plot_api/plot_schema.js]: Oh we could use this for validating `selectedpoints` items as done here:

https://github.com/plotly/plotly.js/blob/a132b85b899dd490216d12823be56ad3861f091f/src/lib/index.js#L497-L499

Comment (etpinard) [reply src/plot_api/plot_schema.js]: ... by _this_ I mean `Lib.isIndex`

File: src/traces/sankey/calc.js [hunk 3]
```diff
@@ -14,13 +14,16 @@ var wrap = require('../../lib/gup').wrap;
 
 function circularityPresent(nodeList, sources, targets) {
 
-    var nodes = nodeList.map(function() {return [];});
+    var nodes = nodeList.map(function() { return []; });
```
Comment (etpinard) [inline src/traces/sankey/calc.js]: Would be nice to reuse:

https://github.com/plotly/plotly.js/blob/a132b85b899dd490216d12823be56ad3861f091f/src/lib/matrix.js#L13-L17

File: test/jasmine/tests/sankey_test.js [hunk 4]
```diff
@@ -58,30 +58,31 @@ describe('sankey tests', function() {
...
-                    link: {
-                        value: [1],
-                        source: [0],
-                        target: [1]
-                    }
-                });
+    describe('No warnings for missing nodes', function() {
+        // we used to warn when some nodes were not used in the links
+        // not doing that anymore, it's not really consistent with
+        // the rest of our data processing.
```
Comment (etpinard) [inline test/jasmine/tests/sankey_test.js]: nicely done :ok_hand:

**Comment by etpinard** [comment]
Great PR :dancer:

**Comment by etpinard** [comment]
... my comments are non-blocking.

File: src/plot_api/plot_schema.js (hunk 1 above)
Comment (alexcjohnson) [reply src/plot_api/plot_schema.js]: Good call -> https://github.com/plotly/plotly.js/pull/2629/commits/cdcd6200907d5caad164933ebbf388bc6532bf19

We probably either have or *should* have this in other places too... `mesh3d` is the other obvious place, but it has bigger issues -> #2630

**Pull request merged**